insert_overwrite (insert+replace) partitions incremental strategy

bryzgaloff opened the pull request:
## Summary
This PR implements insert+replace strategy discussed in #128 which does the following:
* Creates a new staging table with the same structure as the target table.
* Inserts data into the staging table.
* Replaces partitions in the target table from the staging table.

Advantages:
* Only the involved partitions are replaced: this is much cheaper than reinserting the full table which is implemented in other strategies.
* If an insertion fails, the target table is not affected.

## Checklist
Delete items not relevant to your PR:
- [x] Unit and integration tests covering the common scenarios were added
- [x] A human-readable description of the changes was provided to include in CHANGELOG
- [x] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/dbt-clickhouse?pullRequest=201) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/dbt-clickhouse?pullRequest=201) before we can accept your contribution.<br/><hr/>**Anton Bryzgalov** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/dbt-clickhouse?pullRequest=201) it.</sub>

bryzgaloff commented:
At the time of publishing the PR is in WIP (work in progress) state since I need an advice from the community. Thus, neither documentation nor tests are updated.

bryzgaloff commented on dbt/include/clickhouse/macros/materializations/incremental.sql:
```diff
@@ -234,3 +244,36 @@
+          select distinct {{ partition_by|join(', ') }}
+          from {{ intermediate_relation }}
+      {% endset %}
+      {% set partitions_expressions = run_query(select_changed_partitions).rows %}
+    {% else %}
+      {% set partitions_expressions = [] %}
+    {% endif %}
+    {% for partition_expression in partitions_expressions %}
+      {% call statement('replace_partition_' ~ loop.index) %}
+          alter table {{ existing_relation }}
+          replace partition {{ partition_expression }}
```
I am new to DBT and I need an advice for translating output of `run_query` into ClickHouse literal values. Currently `partitions_expressions` is an iterable of [`agate.Row`s](https://agate.readthedocs.io/en/latest/api/columns_and_rows.html#agate.Row). I need to translate `Row{someInt: 1, someStr: 'a', someDate: toDate('2023-11-01')}` to partition expression like `… partition (1, 'a', toDate('2023-11-01')`.

Additionally, we have `partition_by` list of columns names (without types) as an input.

I have several ideas to do it:
* Request target table's columns definitions. Pick the partitioning columns types. Use `CAST('<value-as-string-literal>' AS <target-type>)`.
* Implement translation as a method of `ClickHouseAdapter` and call `adapter.to_partition_expression(…)`. Though I am not sure about the best way to perform translation in adapter: I hope to leverage [`clickhouse-driver`](https://clickhouse-driver.readthedocs.io/en/latest/api.html#clickhouse_driver.Client.execute)'s functionality of translating `params` for this. However, the driver might not have a convenient public interface for the translation.
* Find out some SQL-only way to translate. I.e. compose an SQL query which will do the translation on ClickHouse level: smth like `with partition_expr as (select partitionBy offset {{ loop.index }} limit 1) alter table … replace partition partition_expr …`. This particular example does not work, but maybe there is a possible way.
* I am not sure if `clickhouse-driver`'s usage is discouraged in favour of HTTP interface. If so, a parameterized HTTP query might be an option, though looks the most complex option to me and still requires retrieving partitioning columns' types.

Please let me know if there is an easier dbt-style way to do it. Any advice regarding the suggested options or guidance towards using dbt-specific functions (like requesting a list of table's columns definitions) will be helpful.

I am doing my investigation of the approaches to choose the simplest one. Though I believe the first option is the most reasonable.

@genzgd @simpl1g @tanghyd given that you participated in the original issue #128, could you maybe share some thoughts please?

simpl1g replied:
See `get_partitions` macro in comment above, I believe it should solve this question

bryzgaloff replied:
Very helpful, thank you for your valuable feedback, Konstantin! 🤝 

I have implemented and tested you suggestion with selecting from `system.parts` and it works perfectly fine 👍

bryzgaloff commented on dbt/include/clickhouse/macros/materializations/incremental.sql:
```diff
@@ -41,10 +41,14 @@
 
-  {% elif inserts_only or unique_key is none -%}
+  {% elif
+      inserts_only
+      or unique_key is none
+      and config.get('incremental_strategy', none) != 'insert+replace' -%}
```
A mix of legacy (`insert_only`) and new (`"append"` strategy) approaches to configure the strategy introduces counter-intuitive set of configuration in dbt-clickhouse. To not break the current conditions check flow I have implemented this dirty way of checking, though in general I would suggest refactoring these conditions tree.

To narrow down the reviewed scope, I won't blend the changes into this PR, thus I suggest keeping this dirty-yet-working workaround here.

bryzgaloff commented on dbt/include/clickhouse/macros/materializations/incremental.sql:
```diff
@@ -234,3 +244,36 @@
+    {% else %}
+      {% set partitions_expressions = [] %}
+    {% endif %}
+    {% for partition_expression in partitions_expressions %}
+      {% call statement('replace_partition_' ~ loop.index) %}
+          alter table {{ existing_relation }}
+          replace partition {{ partition_expression }}
```
Also, I have doubts about replacing partitions here in a loop: when multiple partitions are affected, this makes the operation non-atomic: if a subset of partitions is not replaced, the target table might end up being in an inconsistent state. Does dbt have a rollback functionality?

Or does ClickHouse supports manipulating multiple partitions at once? I have not found in documentation and could not compose such a query. Though many Big Data tools (e.g. AWS Athena, Hive) support multiple partitions manipulation when the partitions are listed in a comma-separated fashion.

simpl1g replied:
You can actually do it like this
As an example how to do it in jinja you can [look here](https://github.com/ClickHouse/dbt-clickhouse/blob/b79669ae555f5b060ef42dc14a3d6372cd418b33/dbt/include/clickhouse/macros/persist_docs.sql#L13), it does the same with comments array
```sql
ALTER TABLE ... REPLACE PARTITION 1, REPLACE PARTITION 2, REPLACE PARTITION 3
```

But if you have cluster it will not be atomical. One shard can work and other will fail.

simpl1g commented on dbt/adapters/clickhouse/impl.py:
```diff
@@ -113,7 +113,7 @@ def calculate_incremental_strategy(self, strategy: str) -> str:
             strategy = 'delete_insert' if conn.handle.use_lw_deletes else 'legacy'
         strategy = strategy.replace('+', '_')
-        if strategy not in ['legacy', 'append', 'delete_insert']:
+        if strategy not in ['legacy', 'append', 'delete_insert', 'insert_replace']:
```
I believe `insert_overwrite` is more idiomatic name for this strategy

bryzgaloff replied:
Is it common in dbt plugins domain? I believe insert+replace is more ClickHouse-specific since it describes what the partition does exactly: "**inserts** into a staging table, then **replaces** partitions into a target table"

If it is smth more dbt-like, I agree to rename

simpl1g replied:
It is more dbt specific, for example bigquery or databricks adapters implement it in similar way. You can check example here: https://docs.getdbt.com/reference/resource-configs/bigquery-configs#the-insert_overwrite-strategy

Also this is a good example for README, you can use it to update dbt-clickhouse documentation

bryzgaloff replied:
Hi Konstantin, I have reviewed other plugins and I still insist on `insert+replace` naming due to its unambiguity: the strategy name expresses exactly what it does. And this is an advantage of a suggested naming to me.

However, I rely on the package maintainers' decision here, so please let me know if you force me to change it.

BentsiLeviav replied:
Hi @bryzgaloff 

Thanks for your contribution!
I tend to agree with @simpl1g, as this is the semantic used with dbt (you can read about it [here](https://docs.getdbt.com/docs/build/incremental-strategy)).

I think it is important to align our adapter with the community standards (we can emphasize and detail how this was implemented in the readme file)

bryzgaloff replied:
Ok, updated to `insert_overwrite`, please review the latest version, guys! :)

simpl1g commented on dbt/include/clickhouse/macros/materializations/incremental.sql:
```diff
@@ -72,6 +76,12 @@
         {{ clickhouse__insert_into(target_relation, sql) }}
       {% endcall %}
+    {% elif incremental_strategy == 'insert_replace' %}#}
+      {%- set partition_by = config.get('partition_by') -%}
```
Maybe also better to raise exception if `partition_by` is empty

bryzgaloff replied:
Does not `.get` raise an exception in Jinja if the key is missing? Or does it have regular Python's dict behaviour?

simpl1g commented on dbt/include/clickhouse/macros/materializations/incremental.sql:
```diff
@@ -234,3 +244,36 @@
     {{ drop_relation_if_exists(distributed_new_data_relation) }}
 {% endmacro %}
+
+{% macro clickhouse__incremental_insert_replace(existing_relation, intermediate_relation, partition_by) %}
```
this strategy should also take care about cluster setup and distributed tables. 
We can insert data through Distributed table and then do replace for local table on each shard. This part will be mostly the same as in `clickhouse__incremental_delete_insert` macro

bryzgaloff replied:
Should we maybe defer the implementation till the next iteration? My immediate requirement was to support the strategy for single-node setups (which we are using currently with my team). We may leave a chance to others implement and test the cluster-specific version. I won't have a quick option to test it, thus I won't be able to confirm it works.

Your suggestion with `system.parts` works completely fine, thanks for that! I plan to cover it with tests and document in the next couple of weeks (having a vacation next week :) ).

simpl1g replied:
Not sure about merging without Distributed support, because this is new functionality and mostly experimental, I think this is a question to @genzgd, as he is a maintainer.

But I believe that it should work with cluster from beginning because this is core CH functionality, and most of the production users don't use single server setup

bryzgaloff replied:
We use dbt with ClickHouse without cluster currently, so I have not had an opportunity to test it. I suggest on merging the tested version of my PR now. Once someone needs a clustered version of the strategy, they may contribute it too and test themselves.

If I add the clusters-relates logic, I may not guarantee that it works. However, if the maintainers are ok to rely on their review, I am completely fine with adding a related code snippet.

simpl1g commented on dbt/include/clickhouse/macros/materializations/incremental.sql:
```diff
@@ -234,3 +244,36 @@
+    {%- endcall %}
+    {% if execute %}
+      {% set select_changed_partitions %}
+          select distinct {{ partition_by|join(', ') }}
```
we can do something like this to get partitions
this way we are guaranteed to select all partitions in temp relation
and and will be faster
```sql
{% set partitions = get_partitions(relation) %}
...
{% macro get_partitions(relation) %}
  {% set cluster = adapter.get_clickhouse_cluster_name() %}
  {% set source = 'system.parts' %}
  {% if cluster is not none %}
    {% set source = "cluster('{{ cluster }}', system.parts)" %}
  {% set sql -%}
    SELECT DISTINCT partition_id
    FROM {{ source }} WHERE active AND database = '{{ relation.schema }}' AND table = '{{ relation.identifier }}'
  {%- endset -%}

  {{ return(run_query(sql)) }}
{% endmacro %}

```

bryzgaloff commented:
Hi @simpl1g and @genzgd, I've wrapped up this PR for `insert+replace` strategy implementation: added some fancy documentation in the README and threw in integration tests (taking a page from lw-deletes). Motivation for the feature is given in the PR's description.

Some discussions above are unresolved, waiting on your pearls of wisdom there :)

Please take a peek at the PR when you get a chance, and hit me up with your thoughts. Hoping we can merge the feature soon and let the good times roll with the new strategy! :)

bryzgaloff commented:
Hi @simpl1g and @genzgd! This is a kind reminder about my PR which is ready for your review. **We have successfully battle-tested it internally** ✅ 

We install my version from GitHub currently. It would be nice if you can approve it and release to PyPI. If any adjustments are required, please let me know! 🙏

genzgd commented:
@bryzgaloff I apologize that we haven't yet had the resources to fully review this PR.  As you may have noticed we've been focused on bug fixes and compatibility with the new dbt releases.  Please know that we very much appreciate the contribution (especially with test cases and real world usage) and your work is next on the roadmap as we get time.

If you have a chance to resolve the conflicts over the next few weeks that would be appreciated and make the review just a bit easier.

dev-mkc19 commented on dbt/include/clickhouse/macros/materializations/incremental.sql:
```diff
@@ -234,3 +247,42 @@
+        create table {{ intermediate_relation }} as {{ existing_relation }}
+    {%- endcall %}
+    {% call statement('insert_new_data') -%}
+        insert into {{ intermediate_relation }} select * from {{ new_data_relation }}
```
@bryzgaloff as you said you has a single node deployment.  Here is should be  **{{ on_cluster_clause()}}**. Otherwise error below occurs in my deployment. I'm currently working on adaptation to work on cluster.
```
15:04:39    :HTTPDriver for http://10.100.0.106:8123 returned response code 400)
15:04:39     Code: 36. DB::Exception: Macro 'uuid' and empty arguments of ReplicatedMergeTree are supported only for ON CLUSTER queries with Atomic database engine. (BAD_ARGUMENTS) (version 24.2.1.2248 (official build))
```

Also I have to configure model adding allow_nullable_key option because of otherwise another error occurs. Did you meet with that?
```
14:18:56    :HTTPDriver for http://10.100.0.106:8123 returned response code 400)
14:18:56     Code: 44. DB::Exception: There was an error on [10.100.0.106:9000]: Code: 44. DB::Exception: Partition key contains nullable columns, but merge tree setting `allow_nullable_key` is disabled. (ILLEGAL_COLUMN) (version 24.2.1.2248 (official build)). (ILLEGAL_COLUMN) (version 24.2.1.2248 (official build))
```
```
{{
config(
    materialized = "incremental",
    partition_by = "transaction_date_part",
    incremental_strategy = "insert+replace",
    engine = 'ReplicatedMergeTree',
    order_by = ['city_id', 'transaction_date_part'],
    schema = "analytics",
    settings = {'allow_nullable_key': 1}
)
}}
```

bryzgaloff replied:
Hi @dev-mkc19, thank you for your input! I plan to get back to working on the PR next week. For now, I will keep it without cluster support as I do not have quick-to-setup infrastructure to test it. Feel free to make your own PR adding cluster support 🤝 I may review it to ensure it does not break any of the `insert+replace` semantics. Tag me as a reviewer once it is published!

BentsiLeviav commented:
@bryzgaloff 
Thanks again for your contribution!
I would like to review your PR and merge it within the next few days. Can you sync your fork with the main repo? We upgraded dbt-core to 1.8.0. 

After syncing I'll review this one right away.

bryzgaloff commented:
Hi @BentsiLeviav thank you (and the other reviewers, of course!) for your participation. I am not actively using the plugin right now, but I may get back to handling your review feedback late next week. If the conflicts are not too critical, I might be able to resolve them quickly.

bryzgaloff commented:
Hi @BentsiLeviav I have rebased my contribution onto the latest `main` of this repository. Also, I have renamed the strategy to `insert_replace`. Please have a look! 👋

bryzgaloff commented:
Does this repo has automated tests? I do not have a quick infra to test the contribution, but if there is not testing workflow, I will perform the manual testing.

bryzgaloff commented:
Hi @BentsiLeviav and @genzgd — will you have a chance review and merge the PR soon? I have updated everything according to the review feedback and all the checks have passed. I would like to avoid the need to rebased it again 😅

BentsiLeviav commented:
Hi @bryzgaloff 
Huge thanks for this.
I will review and update you within the next few days.

bryzgaloff commented:
Thank you @BentsiLeviav for the approval! What are the next steps for the PR to be merged?

BentsiLeviav commented:
Hi @bryzgaloff 
Sure, thank you for your contribution!

Before merging this, could you please add to the doc that this feature is experimental, and wasn't tested with cluster setup? 
It is crucial to highlight these 2 points.

Once we are done with that, I'll merge your PR.

Thanks again for your work!

BentsiLeviav commented:
Never mind, I'll take care of it :)

pheepa commented:
@BentsiLeviav @genzgd 

Hi! I’m really interested in contributing to this project, particularly with the cluster functionality.

Before diving into development, I want to ensure I write robust tests for the support of ReplicatedMergeTree and Distributed.

Here’s my plan:

 • Add TestInsertReplaceIncremental (adapted for Replicated) to test_base_incremental.py.
 • Add another version of TestInsertReplaceIncremental (for materialized=distributed_incremental).

Does this approach align with the project’s practices, or would you recommend any adjustments?

Thanks in advance for your guidance!

pheepa commented:
[pr](https://github.com/ClickHouse/dbt-clickhouse/pull/394) here it is!

lddlww commented:
got it, thanks for reply and explanation clearly
